Allow users to estimate query cost before executing it

betodealmeida opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

We currently added to Presto support for estimating the number of bytes scanned (https://github.com/prestosql/presto/pull/806), and we'd like to surface that information to SQL Lab users before they actually run a query.

This PR extends the DB specs with an `allows_cost_estimate` attribute and associated methods, allowing pre-execution costs to be computed from DBs that support it.

In order to use it, the feature flag `ESTIMATE_QUERY_COST` must be enabled, and it needs to be explicitly turned on for each database that supports query cost estimation. When all those conditions are met, a new button will show up in SQL Lab, allowing users to run cost estimates for the whole query or for the selected SQL.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

DBs where the feature is not supported or not enabled are unmodified:

<img width="1698" alt="Screen Shot 2019-09-04 at 8 48 46 AM" src="https://user-images.githubusercontent.com/1534870/64271606-72816180-cef2-11e9-93ea-d379c1fb2ed5.png">

Here's a Presto DB:

<img width="1698" alt="Screen Shot 2019-09-04 at 8 59 15 AM" src="https://user-images.githubusercontent.com/1534870/64271663-917ff380-cef2-11e9-931a-6b32bd843a2f.png">

Waiting for results:

![query_cost](https://user-images.githubusercontent.com/1534870/64650929-3314b280-d3d5-11e9-8346-1f67b6ca9891.gif)

The result:

<img width="1698" alt="Screen Shot 2019-09-04 at 8 58 57 AM" src="https://user-images.githubusercontent.com/1534870/64271692-a066a600-cef2-11e9-8e30-4f4a6bcb6752.png">

And how errors (timeout, syntax errors) are surfaced:

<img width="1698" alt="Screen Shot 2019-09-04 at 8 53 07 AM" src="https://user-images.githubusercontent.com/1534870/64271726-b2e0df80-cef2-11e9-8be8-35c5cce58f46.png">

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

Tested with a Presto cluster that supports query cost estimation, running version 0.319 and with the feature enabled via extra:

```json
{
    "version": "0.319",
    "cost_estimate_enabled": true,
    "metadata_params": {},
    "engine_params": {
       "connect_args": {
           "protocol":"https", 
           "source":"superset"
        }
    },
    "metadata_cache_timeout": {"schema_cache_timeout": 86400, "table_cache_timeout": 86400},
    "default_schemas": ["core", "default"]
}
```

Additionally, I tested:

- Lowering the Presto version removes the button.
- Other DBs don't show the button.
- Turning off the feature flag removes the button.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [X] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

@etr2460 @mistercrunch

betodealmeida committed:
WIP

betodealmeida committed:
Basic functionality working

betodealmeida committed:
Enable per DB

betodealmeida committed:
Show error message

betodealmeida committed:
Small improvements

betodealmeida committed:
Fix lint

etr2460 commented on superset/assets/src/SqlLab/components/EstimateQueryCostButton.jsx:
```diff
@@ -0,0 +1,108 @@
+            <Button
+              bsStyle="warning"
+              bsSize="small"
+              onClick={this.onClick}
```
we probably need some type of indication that this is loading when you click on it, since i'll bet presto doesn't always return super fast.

betodealmeida replied:
It does, it shows a loading wheel.

etr2460 replied:
could you attach a gif to show the behavior?

betodealmeida replied:
Done.

etr2460 commented on superset/db_engine_specs/base.py:
```diff
@@ -643,6 +645,34 @@ def select_star(
         return sql
 
+    @classmethod
+    def estimate_statement_cost(cls, statement, database, cursor, user_name):
```
doc blocks and types please!

betodealmeida replied:
+1

etr2460 commented on superset/db_engine_specs/base.py:
```diff
@@ -643,6 +645,34 @@ def select_star(
+        raise NotImplementedError("Subclasses should implement estimate_query_cost")
+
+    @classmethod
+    def estimate_query_cost(cls, database, schema, sql, source=None):
```
doc blocks and types please!

betodealmeida replied:
+1

etr2460 commented on superset/db_engine_specs/presto.py:
```diff
@@ -373,6 +377,55 @@ def select_star(
         )
 
+    @classmethod
+    def estimate_statement_cost(cls, statement, database, cursor, user_name):
```
doc blocks and types please!

betodealmeida replied:
+1

etr2460 commented on superset/db_engine_specs/presto.py:
```diff
@@ -373,6 +377,55 @@ def select_star(
+            symbols = ["K", "M", "G", "T", "P", "E", "Z", "Y"]
+            while value > 1000 and symbols:
+                prefix = symbols.pop(0)
+                value //= 1000
```
shouldn't this be 1024? Also maybe we should make this magic number a constant

betodealmeida replied:
You're right, I'll fix it.

betodealmeida replied:
Actually, it should be 1000: https://en.wikipedia.org/wiki/Kibibyte

I'll put it in a constant.

betodealmeida replied:
Also, the CPU cost is not in bytes (it's unitless), so we should use 1000.

etr2460 commented on superset/db_engine_specs/presto.py:
```diff
@@ -373,6 +377,55 @@ def select_star(
+        while polled:
+            time.sleep(0.2)
+            polled = cursor.poll()
+        data = db_engine_spec.fetch_data(cursor, 1)
```
can you add a comment showing what this response looks like? It's not super obvious from the code how the result is being pulled out

betodealmeida replied:
Will do!

etr2460 commented on superset/db_engine_specs/presto.py:
```diff
@@ -25,23 +25,27 @@
 
 class PrestoEngineSpec(BaseEngineSpec):
     engine = "presto"
+    allows_cost_estimate = True
```
What version of presto was this added in? Do we need to check the version number to make sure this is available?

betodealmeida replied:
The PR has been merged, but It hasn't been released yet. We patched our Presto cluster with the commit and have it live.

Also, I'm not sure it's trivial to have features depending on the DB version, I remember asking @john-bodley about this for Druid.

etr2460 replied:
I implemented this already in another part of the presto engine spec, relying on the user specifying the version in the extra data field of the db model, but falling back to assuming the most recent version. check out: https://github.com/apache/incubator-superset/blob/master/superset/db_engine_specs/presto.py#L918

betodealmeida replied:
Ah, nice. Let me take a look.

betodealmeida replied:
Cool, I'll remove the DB migration and make it configurable via extras, as well as using the version to see if the DB supports cost estimation.

etr2460 commented on superset/db_engine_specs/presto.py:
```diff
@@ -373,6 +377,55 @@ def select_star(
+            return f"{value} {prefix}{suffix}"
+
+        cost = {}
+        columns = [
```
What happens if the presto format changes? Will we need to make changes here? I know presto likes to add a bunch of breaking changes in new versions...

betodealmeida replied:
Changes won't break this, but might render the response empty. Not sure how to best deal with this.

etr2460 replied:
I guess we'll just need to keep an eye on it

betodealmeida replied:
This is a feature we're expecting to use a lot, so we'll have a good number of eyes. But ideally we'd have integration tests with Presto.

etr2460 commented on superset/migrations/versions/8786d6374caa_add_column_for_query_estimate.py:
```diff
@@ -0,0 +1,40 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Is there a reason we wouldn't want to enable this for all supported dbs if the feature flag is turned on? It seems like the db migration might be overkill here. More of a design question tho.

betodealmeida replied:
I think it might be possible to have users running different versions of Presto (prod and staging, eg), so enabling the feature per DB seems to make sense.

etr2460 replied:
As mentioned above, maybe enable per DB through using the extra data version instead?

betodealmeida replied:
Will do.

etr2460 commented on superset/models/core.py:
```diff
@@ -773,6 +775,10 @@ def name(self):
 
+    @property
+    def allows_cost_estimate(self):
+        return self.db_engine_spec.allows_cost_estimate
```
with the current implementation, shouldn't this be:
`return self.cost_estimate_enabled and self.db_engine_spec.allows_cost_estimate`

betodealmeida replied:
I preferred to implement that logic in the frontend, exposing the 2 separate flags.

etr2460 commented on superset/sql_validators/presto_db.py:
```diff
@@ -157,7 +157,7 @@ def validate(
         with closing(engine.raw_connection()) as conn:
             with closing(conn.cursor()) as cursor:
-                for statement in parsed_query.get_statements():
+                for statement in statements:
```
this could be in a separate PR right? Might be better to pull it out since it's unrelated to this set of work

betodealmeida replied:
Yeah, I was looking at this method and saw that `statements` was already computed. I'll write a separate PR.

etr2460 commented on superset/views/core.py:
```diff
@@ -105,8 +106,12 @@
+
 config = app.config
 CACHE_DEFAULT_TIMEOUT = config.get("CACHE_DEFAULT_TIMEOUT", 0)
+SQLLAB_QUERY_COST_ESTIMATE_TIMEOUT = config.get(
```
not a big fan of configuring different timeouts for each different API call, is there a reason this is needed?

betodealmeida replied:
My impression is that here we want a shorter timeout than usual. The query should return quickly, and if it doesn't return in a few seconds there's probably something wrong and it's not worth waiting for.

etr2460 commented on superset/views/core.py:
```diff
@@ -2391,6 +2396,36 @@ def select_star(self, database_id, table_name, schema=None):
+
+        if len(template_params) > 0:
+            return json_error_response(
+                "Query cost estimation does not support template parameters", status=400
```
why doesn't it support template parameters? Couldn't we just resolve the templates then run the estimation?

betodealmeida replied:
The template processing is tightly coupled with the database (for resolving macros related to latest partition, etc.), and it was hard to bring it on here.

betodealmeida replied:
Actually I got this to work.

etr2460 reviewed (changes requested):
a bunch of comments, also @john-bodley or @villebro should probably double check the db engine specs stuff

betodealmeida committed:
Address comments

betodealmeida commented:
@etr2460, I addressed all the comments. I removed the DB migration, and the feature is enabled per DB in extras. Also, I use the DB version to determine if it's supported.

betodealmeida committed:
Address comments

betodealmeida committed:
Remove logging, fix exception

etr2460 commented on superset/views/core.py:
```diff
@@ -105,8 +106,12 @@
+
 config = app.config
 CACHE_DEFAULT_TIMEOUT = config.get("CACHE_DEFAULT_TIMEOUT", 0)
+SQLLAB_QUERY_COST_ESTIMATE_TIMEOUT = config.get(
```
Can you add this config var to the config file and document what it does there?

betodealmeida replied:
+1

etr2460 commented on superset/views/core.py:
```diff
@@ -2391,6 +2396,34 @@ def select_star(self, database_id, table_name, schema=None):
+    @expose("/estimate_query_cost/<database_id>/<schema>/", methods=["POST"])
+    @event_logger.log_this
+    def estimate_query_cost(self, database_id: int, schema: str = None) -> Response:
+        mydb = db.session.query(models.Database).filter_by(id=database_id).first()
```
i think that `one_or_none()` is a bit clearer than `first()`, but it's just personal preference

betodealmeida replied:
Ah, I wasn't familiar with `one_or_none()`, seems better suited. I'll change.

etr2460 reviewed (approved):
Would you mind making the test plan a little more robust? Test with the feature flag both enabled and disabled, with presto dbs that are configured at a passing version and prior version? With non presto dbs? I'm sure you've tested other cases, but right now the test plan only references the happy path, so a bit more detail would be great.

other than that and my 2 comments here, this lgtm! I'll approve to unblock

betodealmeida committed:
Fix tests

betodealmeida committed:
Bump FAB requirements

betodealmeida committed:
Merge branch 'master' into VIZ-726
